ci-operator/step-registry/ipi/conf/aws/ipi-conf-aws-commands.sh

```diff
@@ -5,58 +5,133 @@ set -o errexit
 set -o pipefail
 
 # TODO: move to image
-curl -L https://github.com/mikefarah/yq/releases/download/3.3.0/yq_linux_amd64 -o /tmp/yq && chmod +x /tmp/yq
+curl -sL https://github.com/mikefarah/yq/releases/download/3.3.0/yq_linux_amd64 -o /tmp/yq && chmod +x /tmp/yq
 
 export AWS_SHARED_CREDENTIALS_FILE="${CLUSTER_PROFILE_DIR}/.awscred"
 
+declare -g MASTER_FAMILY
+declare -g MASTER_TYPE
+
 CONFIG="${SHARED_DIR}/install-config.yaml"
+REGION="${LEASED_RESOURCE}"
 
 expiration_date=$(date -d '8 hours' --iso=minutes --utc)
+existing_zones_setting=$(/tmp/yq r "${CONFIG}" 'controlPlane.platform.aws.zones')
 
 function join_by { local IFS="$1"; shift; echo "$*"; }
 
-REGION="${LEASED_RESOURCE}"
+# Size definition (ipi-conf-aws-ref.yaml): ref.env["SIZE_VARIANT"].documentation
+function get_master_size {
+  master_size=xlarge
+  if [[ "${SIZE_VARIANT:-}" == "xlarge" ]]; then
+    master_size=8xlarge
+  elif [[ "${SIZE_VARIANT:-}" == "large" ]]; then
+    master_size=4xlarge
+  elif [[ "${SIZE_VARIANT:-}" == "compact" ]]; then
+    master_size=2xlarge
+  fi
+  echo ${master_size}
+}
+
+function set_MASTER_TYPE {
+  MASTER_TYPE="${MASTER_FAMILY}.$(get_master_size)"
+}
+
+function choose_preferred_type {
+  local type_preferred
+  local type_backup
+  local zones_cnt_pref
+  local zones_cnt_backup
+
+  type_preferred="${1}.$(get_master_size)"
+  type_backup="${2}.$(get_master_size)"
+
+  if [[ ${existing_zones_setting} == "" ]]; then
+    zones_cnt_pref=$(aws ec2 describe-instance-type-offerings \
+      --region "${REGION}" \
+      --location-type availability-zone \
+      --filters Name=instance-type,Values=\"${type_preferred}\" \
+      --query 'InstanceTypeOfferings[].Location' \
+      --output json | jq -r '. | length')
+    zones_cnt_backup=$(aws ec2 describe-instance-type-offerings \
+      --region "${REGION}" \
+      --location-type availability-zone \
+      --filters Name=instance-type,Values=\"${type_backup}\" \
+      --query 'InstanceTypeOfferings[].Location' \
+      --output json | jq -r '. | length')
+  else
+    zones_expr=$(echo "${existing_zones_setting}" |jq -r '. | join("|")')
+    zones_cnt_pref=$(aws ec2 describe-instance-type-offerings \
+      --region "${REGION}" \
+      --location-type availability-zone \
+      --filters Name=instance-type,Values=\"${type_preferred}\" \
+      --query 'InstanceTypeOfferings[].Location' \
+      --output json \
+      | jq -r ".[] |select(.? | match(\"${zones_expr}\"))" \
+      | wc -l)
+    zones_cnt_backup=$(aws ec2 describe-instance-type-offerings \
+      --region "${REGION}" \
+      --location-type availability-zone \
+      --filters Name=instance-type,Values=\"${type_backup}\" \
+      --query 'InstanceTypeOfferings[].Location' \
+      --output json \
+      | jq -r ".[] |select(.? | match(\"${zones_expr}\"))" \
+      | wc -l)
+  fi
+
+  if [[ ${zones_cnt_pref} -ge ${zones_cnt_backup} ]]; then
```

Comment:
When you switch this logic to check if there are enough zones for the preferred or backup type, then make sure that the zones with availability matches exactly when the zones are explicitly specified in the install config.